### ISSUE TITLE
Defer target metadata resolution to Reporter implementation

### ISSUE THREAD

**Issue by brancz** [elastic/otel-profiling-agent#88]
Currently, there is a strong coupling between the trace handler and what metadata can be attached to the data.

https://github.com/elastic/otel-profiling-agent/blob/0baeb06e93645907253891c254ddf2481941dd3a/tracehandler/tracehandler.go#L135-L138

I'd like to propose making resolving metadata the responsibility of the `Reporter`. This way metadata providers can be extended or work entirely differently from the "upstream" one. In particular, all this would mean is changing the `ReportCountForTrace` to not include all possible metadata in the function signature, but just the PID, the reporter can then call the container metadata provider, APM agent, or whatever else, to retrieve the metadata it wants to attach to the data. The container metadata API already only has the PID in the function signature, so this change would be minimal and natural.

This also has the side effect that the API should be more stable over time since I expect metadata attached to data to be an ever-changing/extended thing.

Very concretely for us this would be required for backward compatibility with Parca Agent, which can use [Prometheus relabeling](https://prometheus.io/docs/prometheus/latest/configuration/configuration/#relabel_config) to optionally attach any metadata discovered from Kubernetes (or other mechanisms). I don't think this mechanism belongs in the otel agent since it's a stronger ecosystem coupling than belongs in an otel project, and fundamentally in otel it's the collector's responsibility to add additional metadata.

@athre0z @florianl @rockdaboot

**Comment by rockdaboot** [comment]
In general: To not do all cleanups at once, I already merged a bunch of smaller PRs. Another one waiting for review is https://github.com/elastic/otel-profiling-agent/pull/84. Taking care of metadata and decoupling it from other packages should be one of the next steps, IMO.

**Comment by brancz** [comment]
Yeah no worries, not a hurry, just wanted to start floating the idea.

**Comment by florianl** [comment]
Hi @brancz 

Before you have opened this issue, I tried to start a discussion internally to reduce the code boiler plate in this repository and move to interfaces were needed. My understanding is, that most users will use this repository for its profiling functionality and not its fully fledged agent capabilities. So your request to use a different format is valid, I think and supports the move to proper interfaces. With https://github.com/elastic/otel-profiling-agent/issues/89 exists an issue around the packages in this repository, that are not directly related to profiling and could be removed.

**Comment by brancz** [comment]
Thank you for starting that! Like I already commented there, all for this, but since this is part of otel it should still be minimally useful and integrating into the otel ecosystem (I believe that would be significantly less code to maintain than it is today). Otherwise very happy with this!

**Comment by brancz** [comment]
This is now resolved with #90 and https://github.com/elastic/otel-profiling-agent/commit/bda1135f6c3a50a61e06319a07327accd8aedbdf.
